oak_restricted_kernel/src/syscall/key.rs

```diff
@@ -16,20 +16,26 @@
 
 use crate::snp_guest::DerivedKey;
 
-use super::fd::FileDescriptor;
+use super::fd::{copy_max_slice, FileDescriptor};
 use alloc::boxed::Box;
-use core::cmp::min;
 use oak_restricted_kernel_interface::{Errno, DERIVED_KEY_FD};
 
 #[derive(Default)]
 struct DerivedKeyDescriptor {
     key: DerivedKey,
+    index: usize,
+}
+
+impl DerivedKeyDescriptor {
+    fn new(key: DerivedKey) -> Self {
+        Self { index: 0, key }
+    }
 }
 
 impl FileDescriptor for DerivedKeyDescriptor {
     fn read(&mut self, buf: &mut [u8]) -> Result<isize, oak_restricted_kernel_interface::Errno> {
-        let length = min(self.key.len(), buf.len());
-        buf.copy_from_slice(&self.key[..length]);
+        let length = copy_max_slice(&self.key[self.index..], buf);
+        self.index += length;
         Ok(length as isize)
```

Comment:
unrelated to this PR,but I think we should be returning an usize here. There's no reason to use a signed integer here, the length should never be negative.

Reply:
Errors can be negative.

Reply:
oh noes, gotcha. So we're using the negative values to designate errors. And we also use the rust Result type, but errors are just represented as negative numbers. 

I assume at the end the application just gets the number. 

I understand, but it remains misleading imo. 

eg `Ok(Errno::EINVAL)` is a valid return type atm.

Reply:
thanks for explaining. Let's try and copy that explanation in there as a comment somewhere.